From c0ad9515060dd306144e719e4e1eb781ed7dbda3 Mon Sep 17 00:00:00 2001
From: Fredrik Orderud <fredrik.orderud@ge.com>
Date: Wed, 27 Nov 2019 20:27:19 +0100
Subject: [PATCH] Disable TextureSwizzle when targeting WASM

The WebGL 2.0 specification explicitly does not support texture swizzles. Done
to fix "WebGL: INVALID_ENUM: texParameter: invalid parameter name" when running
in Chrome or Firefox.

Change-Id: Ic7e22e0f623095245274924095cb63fd0ff7e8c2
Reference: https://www.khronos.org/registry/webgl/specs/latest/2.0/#5.19
Fixes: QTBUG-80287
---
 src/gui/opengl/qopenglfunctions.cpp | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/src/gui/opengl/qopenglfunctions.cpp b/src/gui/opengl/qopenglfunctions.cpp
index 8ec814296a..44e42e1560 100644
--- a/src/gui/opengl/qopenglfunctions.cpp
+++ b/src/gui/opengl/qopenglfunctions.cpp
@@ -388,8 +388,12 @@ static int qt_gl_resolve_extensions()
                 | QOpenGLExtensions::MapBufferRange
                 | QOpenGLExtensions::FramebufferBlit
                 | QOpenGLExtensions::FramebufferMultisample
-                | QOpenGLExtensions::Sized8Formats
-                | QOpenGLExtensions::TextureSwizzle;
+                | QOpenGLExtensions::Sized8Formats;
+#ifndef Q_OS_WASM
+            // WebGL 2.0 specification explicitly does not support texture swizzles
+            // https://www.khronos.org/registry/webgl/specs/latest/2.0/#5.19
+            extensions |= QOpenGLExtensions::TextureSwizzle;
+#endif
         } else {
             // Recognize features by extension name.
             if (extensionMatcher.match("GL_OES_packed_depth_stencil"))
-- 
2.22.0.windows.1

